### PR TITLE
Grinders now use radials

### PR DESCRIPTION
**Pull request by nicbn**
[Changelogs]: #
:cl:
add: All-In-One Blender UI uses a radial menu now. You can see the contents by examining.
/:cl:

If there are no reagents and beaker is empty or there is no beaker, clicking it will eject the contents.
image
image

Images outdated, eject button is always on top

### CODE REVIEW

**Comment by MACIEKBAKI** [comment]
^^ i have to say thats a design flaw.

**Comment by nicbn** [comment]
I'll make it so that examining it will show the reagents

**Comment by vuonojenmustaturska** [comment]
why should it show you the reagents though? grinders are dumb machines

**Comment by nicbn** [comment]
I'll make it so that examining it will show the reagents if you are wearing glasses that let you examine beakers

File: code/_onclick/hud/radial.dm [hunk 1]
```diff
@@ -250,25 +250,28 @@ GLOBAL_LIST_EMPTY(radial_menus)
 	if(current_user)
 		current_user.images -= menu_holder
 
-/datum/radial_menu/proc/wait()
+/datum/radial_menu/proc/wait(atom/user = null, atom/anchor = null, require_near = FALSE)
```
Comment (Cyberboss) [inline code/_onclick/hud/radial.dm]:
```suggestion
/datum/radial_menu/proc/wait(atom/user, atom/anchor, require_near = FALSE)
```

File: code/_onclick/hud/radial.dm [hunk 2]
```diff
@@ -250,25 +250,28 @@ GLOBAL_LIST_EMPTY(radial_menus)
...
 	while (current_user && !finished && !selected_choice)
 		if(custom_check_callback && next_check < world.time)
 			if(!custom_check_callback.Invoke())
 				return
 			else
 				next_check = world.time + check_delay
+		if(require_near && user && anchor)
+			if(!in_range(anchor, user))
```
Comment (Cyberboss) [inline code/_onclick/hud/radial.dm]: Condense the two if statements

**Comment by Barhandar** [comment]
Does AI have said glasses?

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by nicbn** [comment]
AI can't even examine things

**Comment by Barhandar** [comment]
I've seen AIs request beaker be popped into blender to see the reagents inside, and it's not your problem whether they're supposed to interact with anything.

**Comment by nicbn** [comment]
Ok I'm going to make it so that when AI interacts with the blender it will have the option of examining it via a button.

**Comment by ExcessiveUseOfCobblestone** [comment]
We have goggles, machines, PDAs, research items, etc. explicitly for determining the contents of a reagent container and/or manipulating those contents. Having holes for people to sabotage/conceal are **good**.

**Comment by Vidiveni** [comment]
This just adds tedium. I hate when buttons move around Windows 10 style. Not being able to see finished reagents without extra "examine" clicks will be very annoying, especially for cooks, botanists and miners.

**Comment by ExcessiveUseOfCobblestone** [comment]
AI can only examine relative to their core...

Also the point is botanists,chefs,etc. should go to the proper equipment or wear the proper gear to determine reagents or, heaven forbid, they have to put trust in the items they're grinding.

**Comment by nicbn** [comment]
Ok let's break everything down

>SEE THE REAGENTS IN A BEAKER IN A PINCH

If you are wearing science googles / beer googles / reagent googles just shift click, if you are AI just click the icon in the radial

>I hate when buttons move around Windows 10 style

Currently this pops up a window, you can't even be sure where. Also the window may lag

>Not being able to see finished reagents without extra "examine" clicks 

Before: get near, click, wait
After: hold shift, click (if you have proper gear)

**Comment by nicbn** [comment]
>Can you at least give the buttons static positions? So eject doesn't shift to the left sometimes.

This is possible but just not worth it. The only time eject shifts is after blending and after inserting the first object in, so it shouldn't be much of an issue. Trust me, it's much better than having the eject button appear in a popup window.

**Comment by nichlas0010** [comment]
@zxaber just press the black cross-thingy in the middle

**Comment by nicbn** [comment]
Moving away also dismisses the menu
I'll make it so that eject will be always in the top for mobs

**Comment by miniusAreas** [comment]
i feel like radials are going down the windows 8 line, can we get both radials and the previous options?

**Comment by ExcessiveUseOfCobblestone** [comment]
No because then you have to constantly maintain both UIs simultaneously per item.

**Comment by optimumtact** [comment]
I really think in the interests of convention we should not mix and match interface types and most machines are going to have a browser based menu. :thinking: 

In summary I'm really not sure on this, would like to see more opinions if people have them.

**Comment by kriskog** [comment]
My initial reaction is that I don't want it. I can imagine it being useful because you don't need to open the window and have your inputs held hostage. At the same time, the radials means your interaction with the machine will be less precise. Also, how do they work in darkness/bad lighting?

That said, I didn't think the RCD would benefit from radials and boy was I wrong. But that did just replace cycling.

**Comment by ShizCalev** [comment]
>add: All-In-One Blender UI uses a radial menu now. If you have proper gear, you can see the contents by examining.
> balance: You need gear that lets you examine beakers to see reagents if you want to see reagents inside of a blender.


> be chef
> start grinding food for nutriment
> can't remember if you put two or three donkpockets into the grinder
> have to click eject because you're not wearing science goggles to see inside the glass blender
> decide not to click it and just grind away
> can't tell what's inside the glass container now because you're still not wearing science goggles
> eject

This change affects hydroponics, xenobiology, the bartender, the chef, chemistry, and virology.

I get the not being able to see reagents thing, that makes sense as that's how the rest of the examine system works - but not being able to see how many donkpockets are inside the blender without science goggles on is a hard pass from me.

**Comment by nicbn** [comment]
>can't remember if you put two or three donkpockets into the grinder

You only need glasses to see LIQUID contents in the beaker
You can always see SOLID contents like donk pockets

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm [hunk 3]
```diff
@@ -125,104 +126,85 @@
...
-	popup.set_content(dat)
-	popup.set_title_image(user.browse_rsc_icon(icon, icon_state))
-	popup.open(TRUE)
-	return
 
-/obj/machinery/reagentgrinder/Topic(href, href_list)
-	if(..())
-		return
-	var/mob/user = usr
-	if(!user.canUseTopic(src))
```
Comment (ShizCalev) [inline code/modules/reagents/chemistry/machinery/reagentgrinder.dm]: might be overlooking something, but are there no user stat checks anywhere in this?

Comment (nicbn) [reply code/modules/reagents/chemistry/machinery/reagentgrinder.dm]: `ui_interact` already handles this

Comment (ShizCalev) [reply code/modules/reagents/chemistry/machinery/reagentgrinder.dm]: no it doesn't.

Comment (ShizCalev) [reply code/modules/reagents/chemistry/machinery/reagentgrinder.dm]: likewise if you're handcuffed with it still open

**Comment by kriskog** [comment]
To add to my previous comment, seems the solid contents are returned to chat. Iirc, grinders hold up to 13 items, which means 13 lines + the examine text in chat if I examine them when full? Add 3-4 reagents in beaker, and we got a ~20 line examine block. Chat is already moving fast, I would rather keep my chat clean and use the window tbh. 

As for hiding reagents in beaker, every department that has roundstart grinders has separate machinery for splitting reagents. It doesn't make the information inaccessible, just more tedious.

**Comment by ExcessiveUseOfCobblestone** [comment]
Does the chat not stack for this? I sincerely doubt you're going to have 13 items of different names in a blender at a given time UNLESS RPG loot event.

**Comment by ExcessiveUseOfCobblestone** [comment]
Autolathes could have a radial opening to different categories then a window upon selecting which, I don't think that's a stretch.

File: code/modules/reagents/chemistry/machinery/reagentgrinder.dm (hunk 3 above)
Comment (nicbn) [reply code/modules/reagents/chemistry/machinery/reagentgrinder.dm]: Fixed

**Comment by nicbn** [comment]
Stop complaining about tedious
No hugbox for y'all in this PR

**Comment by PKPenguin321** [comment]
Would it be so hard to add an extra button that tells the reagent contents in the chat?

**Comment by nicbn** [comment]
Just shift click or right click then examine

**Comment by PKPenguin321** [comment]
Oh right good idea. It could have an examine message like
```
The screen displays that the current beaker holds:
10u of X
15u of Y
```
But maybe you'd have to be adjacent to see it. Good idea, I love the fluidity of the new UI and not having to sacrifice any functionality is excellent

**Comment by nicbn** [comment]
Ok I'll be changing it so that science googles aren't required to view reagents

**Comment by nicbn** [comment]
A major change was made, so I'm opening a new PR.

**Pull request closed without merging**